api/src/main/java/org/apache/iceberg/expressions/BoundExtract.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import java.util.List;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.variants.Variant;
+import org.apache.iceberg.variants.VariantDataUtil;
+
+public class BoundExtract<T> implements BoundTerm<T> {
+  private final BoundReference<Variant> ref;
+  private final String path;
+  private final List<String> fields;
+  private final String fullFieldName;
+  private final Type type;
+
+  BoundExtract(BoundReference<Variant> ref, String path, Type type) {
+    this.ref = ref;
+    this.path = path;
+    this.fields = VariantDataUtil.parsePath(path);
+    this.fullFieldName = Joiner.on(".").join(fields);
```

Comment:
I'm a little concerned about this.  This appears to tie the extraction to a specific field, but that's somewhat limiting.  For example, JsonPath allows for extracting multiple values with an expression and the result is always a list of matched results.  While we can narrow our implementation, I think that's going to be very limiting.

Reply:
This will depend on the representation that we choose for storing the lower and upper bounds. Right now, we are thinking that it will be conceptually a map of field name to value (even if possibly not represented as an object). That would require converting multi-part names to strings, which we typically do using a simple joiner like this. If there are collisions, we would omit the values.